packages/labs/react/src/create-component.ts

```diff
@@ -4,73 +4,87 @@
  * SPDX-License-Identifier: BSD-3-Clause
  */
 
+import type React from 'react';
+
 const NODE_MODE = false;
+const DEV_MODE = true;
 
-import type React from 'react';
+/**
+ * Creates a type to be used for the props of a web component used directly in
+ * React JSX.
+ *
+ * Example:
+ *
+ * ```ts
+ * declare module "react" {
+ *   namespace JSX {
+ *     interface IntrinsicElements {
+ *       'x-foo': WebComponentProps<XFoo>;
+ *     }
+ *   }
+ * }
+ * ```
+ */
+export type WebComponentProps<I extends HTMLElement> = React.HTMLAttributes<I> &
+  // TODO(augustjk) Consider omitting keyof LitElement to remove "internal"
+  // lifecycle methods or allow user to explicitly set the prop type instead
+  // of grabbing from class
+  Partial<Omit<I, keyof HTMLElement>>;
 
-// Match a prop name to a typed event callback by
-// adding an Event type as an expected property on a string.
+/**
+ * Type of the React component wrapping the web component. This is the return
+ * type of `createComponent`.
+ */
+export type ReactWebComponent<
+  I extends HTMLElement,
+  E extends EventNames = {}
+> = React.ForwardRefExoticComponent<
+  React.PropsWithoutRef<WebComponentProps<I> & EventListeners<E>> &
+    React.RefAttributes<I>
+>;
+
+/**
+ * Type used to cast an event name with an event type when providing the
+ * `events` option to `createComponent` for better typing of the event handler
+ * prop.
+ *
+ * Example:
+ *
+ * ```ts
+ * const FooComponent = createCompoennt({
+ *   ...
+ *   events: {
+ *     onfoo: 'foo' as EventName<FooEvent>,
+ *   }
+ * });
+ * ```
+ *
+ * `onfoo` prop will have the type `(e: FooEvent) => void`.
+ */
 export type EventName<T extends Event = Event> = string & {
   __event_type: T;
```

Comment:
while you're here, we'd normally write this as `__eventType`, ie the `__` is a prefix, the whole identifier is not snake_case.